src/intellisense/langIntellisenseProvider.ts

```diff
@@ -646,21 +646,44 @@ class ReferenceCompletionProvider extends IntellisenseProvider implements vscode
                     return labels.map(label => new vscode.CompletionItem(label, vscode.CompletionItemKind.Reference));
                 }).flat();
             case 1:
-                const bibUri = vfs.pathToUri(`${OUTPUT_FOLDER_NAME}/output.bbl`);
-                const content = new TextDecoder().decode( await vfs.openFile(bibUri) );
-                const regex = /\\bibitem\{([^\}]*)\}/g;
                 const items = new Array<vscode.CompletionItem>();
-                let match: RegExpExecArray | null;
-                while (match = regex.exec(content)) {
-                    const item = new vscode.CompletionItem(match[1], vscode.CompletionItemKind.Reference);
-                    items.push(item);
+                items.push(... await this.getBibCompletionItems(vfs, '/'));
+                if ('output.bbl' in await vfs.list(vfs.pathToUri(OUTPUT_FOLDER_NAME))) {
```

Comment:
why use `vfs.list`, lack of efficiency.

Reply:
I cannot tell why you use a separate "get bib completion items" function. So what is the remaining part for? "extra items"? Do you imply the items are extra or something?

I provide you some possible options to make the code more readable:
- use one standalone function (like your added `getBibCompletionItems`) to provide all bib completions
- move the remaining part to another standalone function, with a meaningful name

Reply:
> why use `vfs.list`, lack of efficiency.

Since I have to find the bib file name in the project, it may require a traverse on the file name. As for the `vfs.list`, I found `const rootFiles = await vfs.list(vfs.pathToUri('/'));` in `CommandCompletionProvider.reloadCommands` which seems to match the requirement.



> "get bib completion items" function. So what is the remaining part for? "extra items"? Do you imply the items are extra or something?
> 
> I provide you some possible options to make the code more readable:
> 
> * use one standalone function (like your added `getBibCompletionItems`) to provide all bib completions
> * move the remaining part to another standalone function, with a meaningful name

Then I might parse it into independent part of function named as 'getBibCompletionItemsFromOutput' and also refactor the `getBibCompletionItems` into `getBibCompletionItemsFromBib`. The remaining part is simply original one which get bibitems from `bbl`.

Reply:
You need to auto-complete the "bib items" in the included "bib file" in the *MAIN DOCUMENT*, but not all the bib files.
And you should not expect all bib files are located at the root

![](https://github.com/iamhyc/Overleaf-Workshop/assets/9068301/f2ef7fed-e708-4cd5-963f-eedfefeae153)

Reply:
> You need to auto-complete the "bib items" in the included "bib file" in the _MAIN DOCUMENT_, but not all the bib files. And you should not expect all bib files are located at the root
> 
> ![](https://private-user-images.githubusercontent.com/9068301/294750704-f2ef7fed-e708-4cd5-963f-eedfefeae153.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ2MjIyMTUsIm5iZiI6MTcwNDYyMTkxNSwicGF0aCI6Ii85MDY4MzAxLzI5NDc1MDcwNC1mMmVmN2ZlZC1lNzA4LTRjZDUtOTYzZi1lZWRmZWZlYWUxNTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMDdUMTAwNTE1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YmZjZmUxMDNmMWEyNmRiMmE0YTQzNzMxNGU0ZThkZWI5ZGNmN2QwODNmZjY2MjRmOWEzNTk3ZTIxMDYyZWIwOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8Nt-OL0oLvHsejYarIsc6kT-H4u21St5RDXLTInocEU)

This implementation will traverse all the files in project, thus actually it will search out those not in the roor file; As for the previous part, I might implement it by searching the main document soon.

Reply:
Ok, my bad, but please *DO NOT* simply traverse all files, it is time-expensive and not a expected behavior.

Please turn this PR into draft now and let me know when you are ready for review.

Reply:
Sure, I'll close this PR in a minute.

Reply:
ok, you don't get my point, so I do it for you.

FYI, the draft button is on the left top of the page.

Let me know when you are ready for review by clicking the button, please.

![](https://github.com/iamhyc/Overleaf-Workshop/assets/9068301/61f7ac63-c152-48a4-b56e-6ec0c93c52e1)

Reply:
A draft PR will not annoy anyone when you push the commits to the branch. And you don't have to introduce extra closed/reopen status.